### PR TITLE
Feature/choice set manager

### LINKED ISSUE #123: [SDL 0272] JavaScript Suite SDL Manager
Proposal: [JavaScript Suite SDL Manager](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0272-sdl-javascript-manager-layer.md)

> In a similar vein to [this proposal](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0171-android-manager-apis.md), the goal is to introduce a manager layer to the JavaScript Suite, closely matching the Java Suite and iOS manager APIs. These manager classes are designed to abstract out complex flows and states for common tasks, such as file handling and displaying information to the screen. 

Review: https://github.com/smartdevicelink/sdl_evolution/issues/901

Steering Committee Decision:

> The Steering Committee voted to accept this proposal with the revisions listed below:

- (Item 2) Pass an AppConfig instance into the SDLManager constructor.  The SDLManager `start()` method can perform the basic validation checks prior to starting the LifecycleManager to ensure the minimum properties of an AppConfig are provided, as well as a ManagerListener being defined.
- (Item 3) Single-method listeners to be replaced with simple JavaScript-oriented callback functions.
- (Item 7) Promises should be returned by manager methods which ultimately send one or more RPCs resulting in a single response. Developers then have the freedom to decide if/how they would like to handle the resolution or rejection of the RPC response. 
- (Item 12) Remove `getInternalInterface()` in favor of utilizing the reference to the LifecycleManager

> The proposal .md file was updated to reflect these revisions on 3/3/2020.

### PR DESCRIPTION
**Pull request by crokita**
Partially fixes #123 

### Risk
This PR makes minor API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
Unit tests added that match the tests added in the Java Suite library for the choice set related classes

### Summary
Adds the choice set manager which feature matches the Java Suite library for all the screen manager APIs related to the choice set manager.

Note that presentSearchableChoiceSet would not be supported in the latest develop Generic HMI, and that OnKeyboardInput messages haven't been received.

### CODE REVIEW

File: lib/js/src/manager/screen/choiceset/ChoiceCell.js
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/ChoiceCell.js:226]: The .equals() method is part of the SdlFile class, not the SdlArtwork class. Because of that, it won't check the _isTemplate property which is part of the SdlArtwork class.

File: lib/js/src/manager/screen/choiceset/_CheckChoiceVrOptionalOperation.js
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_CheckChoiceVrOptionalOperation.js:82]: You're using await here but that was already done when deletePromise was initialized earlier.

File: lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js:178]:
```suggestion
        if (this._pendingPresentationSet !== null && this._pendingPresentationSet.getChoices() !== null) { // type is ChoiceSet
```

File: lib/js/src/manager/screen/choiceset/_PresentKeyboardOperation.js
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_PresentKeyboardOperation.js:61]:
```suggestion
        this._cancelId = cancelId;
        this._keyboardProperties = originalKeyboardProperties;
```

File: lib/js/src/manager/screen/choiceset/_PresentKeyboardOperation.js
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_PresentKeyboardOperation.js:200]:
```suggestion
            console.error(`Canceled the keyboard set ${response.getSuccess() === Result.SUCCESS ? 'successfully' : 'unsuccessfully'}`);
```

**Review by renonick87 — CHANGES_REQUESTED** [review]
See requested feedback and remember to run `npm run lint` to resolve any issues in documentation.

File: lib/js/src/manager/screen/choiceset/ChoiceCell.js
Comment (crokita) [reply lib/js/src/manager/screen/choiceset/ChoiceCell.js:226]: I believe this is the way the Java version does this. Their way of comparing object equality is by implementing the `hashCode`/`equals` method. The SdlArtwork class doesn't implement this but the parent SdlFile does. Unless this is a bug with the Java code as well, it seems to ignore the extra properties SdlArtwork offers as part of the equality check.

**Comment by crokita** [comment]
PR updated. I didn't include unrelated linter fixes in this PR: that can be done in a separate PR. Also note:
- I moved systemcapabilitylistener logic to the _ChoiceSetManagerBase to fix an issue with _displayName never being set by looking at the display capabilities
- _ChoiceSetSelectionListener and _KeyboardListener are now public classes
- Fixed ChoiceCell's equals() method to not error out when comparing null voice commands

File: lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js
Comment (renonick87) [inline lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js:532]: This method looks to be the same as the `_SubManagerBase._handleDisplayCapabilities` method.

File: tests/managers/index.js
Comment (renonick87) [inline tests/managers/index.js:40]: Something with the most recent update is causing failures in the `voiceCommandManagerTests` and `voiceCommandUpdateOperationTests` for me. I've tested this with manticore across multiple browsers. Everything passes if these tests are moved above the `textAndGraphicManagerTests` and run first though.

File: lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js
Comment (crokita) [reply lib/js/src/manager/screen/choiceset/_ChoiceSetManagerBase.js:532]: Line 538 has the small difference. The display name from the system capabilities needs to be taken and saved as displayName in this class

File: tests/managers/index.js
Comment (crokita) [reply tests/managers/index.js:40]: That's annoying. I have moved those tests and got the same results now.

**Comment by crokita** [comment]
Updated. See comments in response to feedback

**Review by renonick87 — APPROVED** [review]
Since this PR contains updates from https://github.com/smartdevicelink/sdl_javascript_suite/pull/357, merging the ChoiceSetManager will have to wait until after that PR is merged first.

**Comment by renonick87** [comment]
Updates from https://github.com/smartdevicelink/sdl_javascript_suite/pull/357 have been reverted from this PR. This PR is set to be merged.

**Comment by crokita** [comment]
Squashing to avoid accidentally clearing out this PR: #357

**Pull request merged**
